### ISSUE TITLE
Changed Auto-remove settings, all previous backups lost with each new one made

### ISSUE THREAD

**Issue by moved-to-gitlab** [bit-team/backintime#733]
Using the native version of backintime on Ubuntu 16.04 Desktop I have a new problem today. I had about one year's worth of backups and lately they were saving with a (WITH ERRORS) message at the end of the name. Looking at the [last log](https://gist.github.com/moved-to-gitlab/5d21398e431409a4bc4e99ff260d9d50), there are no errors. Looking at the results of `zgrep "backintime" /var/log/syslog*` [shown here](https://gist.github.com/moved-to-gitlab/8f17dd36d62551a3eb32b8997155bd1e), I don't see what went wrong.

Right now backintime runs every two hours as usual, but each new backup removes all previous backups. It only keeps the most recent one, everything else is gone. Before it kept many, many old versions.

Prior to all backups being lost, I increased the Auto-remove settings for "if free space is less than" from 20GiB to 50Gib and also increased the "If free inodes is less than 10%" from 5% to 10%. I did this because the file the original "(WITH ERRORS)" message was referring to was a large Thunderbird archive that was (probably) over 20GiB.

I have backups of my data made with another backup tool so I haven't lost any essential data, but I would like to recover my backintime snapshots and learn what went wrong here. Any advice for further troubleshooting please?

EDIT: I've changed the 'free inodes' auto-remove back to 5% and newly made snapshots are once again being saved, However my old ones are still gone and this is a 2TB drive used solely for backups. It is nowhere near 5% or even 50% full. I'm backing up less than 250GB in total so a settings change from 5% to 10% should make no difference, but it just did.

**Comment by colinl** [comment]
How much free space is there now?

Colin

**Comment by moved-to-gitlab** [comment]
1.5Tb free space available now, it's a 2Tb drive.
```
df -h
/dev/dm-4                    1.8T  234G  1.5T  14% /Backups/backintime
```

**Comment by Germar** [comment]
> /var/log/syslog:Apr  2 08:30:31 desktop backintime (user/1): INFO: Keep min free disk space: 51200 MiB
> /var/log/syslog:Apr  2 08:30:31 desktop backintime (user/1): INFO: free disk space: 39976 MiB


You had only 39GiB free space when BiT started to delete snapshots.

**Comment by moved-to-gitlab** [comment]
> You had only 39GiB free space when BiT started to delete snapshots.

Yes, that's correct. 39Gib left on a 2TB drive. Now I have over 1.5TB free space and all backups from the past year (excepting today's) deleted. This was not the expected result.

I would expect backintime to free up about the same amount of used space as it needed for the next backup, and not delete everything on the entire drive. That's what it used to do before I made a minor change to the 'free inodes' setting, to troubleshoot the backups that were saving with errors.

TL;DR 'free inodes' auto-remove settings change from 5% to 10% wipes all existing backups on a 2TB drive each time backintime runs. Only the latest (~200GB) backup is ever saved. I don't believe that's the expected result, right?

**Comment by Germar** [comment]
If there is less free space than you configured BiT will delete the oldest snapshot in the row and check free space. If there is still less free space it will again delete the oldest snapshot until there is only one snapshot left.

From your syslog I can see, that free space did not increase during deleting snapshots (which is curious, was there an other process filling up your drive at the same time? ). When BiT stopped deleting snapshots (because only one was left) there was only 39628MiB free space. `Free inodes` had nothing to do with deleting those snapshots. At least not in the 8:25 snapshot

**Comment by moved-to-gitlab** [comment]
Everything you mention makes perfect sense and that's how BiT used to function. When I noticed the snapshot save errors, I saw the file it was having trouble with was about the same size as the free space setting for the auto-remove function, so I changed this (plus the inodes percentage) a little, to see if that was the cause.

This is a separate physical drive in a desktop computer with multiple drives. The backup target drive is solely used by BiT for backups. Nothing else uses it and until this morning everything was working well. Very reliable and trouble-free for about a year of daily usage, with backups every two hours.

The backup drive is ext4, LUKS encrypted and mounts automatically at boot. That's probably irrelevant but I mention it just in case.

**Comment by moved-to-gitlab** [comment]
I've just tested this again and I can reproduce the error by changing the Settings > If free inodes is less than: > 5%, back up to 10%. For a 2TB drive making a ~200GB backup, it once again deleted all the backups it had made today. Only the newest one is saved.

As I mentioned earlier I use more than one tool to make backups and have not lost any important data, but BiT is my preferred tool for this and I'm concerned that it is not producing the expected results and deleting data. Is my understanding of "free inodes" incorrect? This is a large drive with loads of free space, and I cannot understand why all data is lost when making a small settings change.

Here is the newest output of `zgrep "backintime" /var/log/syslog*` from today only. : https://gist.github.com/moved-to-gitlab/8814ed9bd1357b6644879fd8017699a1

**Comment by Germar** [comment]
> /var/log/syslog:Apr  2 21:45:45 desktop backintime (user/1): INFO: Keep min free disk space: 51200 MiB
> /var/log/syslog:Apr  2 21:45:45 desktop backintime (user/1): INFO: free disk space: 37937 MiB
> /var/log/syslog:Apr  2 21:46:16 desktop backintime message repeated 4 times: [ (user/1): INFO: free disk space: 37937 MiB]
> /var/log/syslog:Apr  2 21:46:18 desktop backintime (user/1): INFO: Keep min 10% free inodes
> /var/log/syslog:Apr 2 21:46:20 desktop backintime (user/1): INFO: Unlock

In this snapshot BiT again deleted five snapshots only because there was only 37937MiB free space. The message `INFO: Keep min 10% free inodes` only said that BiT now check if there is at least 10% free inodes. But that rule didn't delete any snapshot. If it would have deleted snapshots because of the `Min free inodes` Rule it would have printed a line `INFO: free inodes: XX%`

Please post the output of
```
df -h $(sed -n -e 's/^profile1.snapshots.path=\(.*\)/\1/p' ~/.config/backintime/config)
df -hi $(sed -n -e 's/^profile1.snapshots.path=\(.*\)/\1/p' ~/.config/backintime/config)
```

**Comment by moved-to-gitlab** [comment]
Here's the output:
```
user@desktop:~$ df -h $(sed -n -e 's/^profile1.snapshots.path=\(.*\)/\1/p' ~/.config/backintime/config)
Filesystem                   Size  Used Avail Use% Mounted on
/dev/mapper/ubuntu--vg-root  221G  171G   39G  82% /
```
```
user@desktop:~$ df -hi $(sed -n -e 's/^profile1.snapshots.path=\(.*\)/\1/p' ~/.config/backintime/config)
Filesystem                  Inodes IUsed IFree IUse% Mounted on
/dev/mapper/ubuntu--vg-root    15M  505K   14M    4% /
```
These values are correct, however they are for the root filesystem and do not reflect all the empty space on the physically separate backup drive (and which mounts at `/Backups/backintime` at boot)

Here's everything for the internal drives:
```
user@desktop:~$ df -h
Filesystem                   Size  Used Avail Use% Mounted on
udev                         7.8G     0  7.8G   0% /dev
tmpfs                        1.6G  9.7M  1.6G   1% /run
/dev/mapper/ubuntu--vg-root  221G  170G   41G  81% /
tmpfs                        7.9G  316K  7.9G   1% /dev/shm
tmpfs                        5.0M  4.0K  5.0M   1% /run/lock
tmpfs                        7.9G     0  7.9G   0% /sys/fs/cgroup
/dev/sdb1                    110G   89G   17G  85% /mnt/Music
/dev/sdd5                    550G  373G  150G  72% /Backups/default
/dev/sda1                    472M  236M  212M  53% /boot
/dev/dm-4                    1.8T  234G  1.5T  14% /Backups/backintime
tmpfs                        1.6G  124K  1.6G   1% /run/user/1000
/home/user/.Private           221G  170G   41G  81% /home/user
```
And here's `/etc/fstab`:
```
user@desktop:~$ cat /etc/fstab
# <file system> <mount point>   <type>  <options>       <dump>  <pass>
/dev/mapper/ubuntu--vg-root /               ext4    errors=remount-ro 0       1
# /boot was on /dev/sda1 during installation
UUID=69217302-de5d-47e4-9783-c46077a296f7 /boot           ext2    defaults        0       2
#/dev/mapper/ubuntu--vg-swap_1 none            swap    sw              0       0
/dev/mapper/cryptswap1 none swap sw 0 0
# Backups
UUID=5b88825c-20b4-4a46-9e15-c98372723780       /mnt/backups/default    ext4    defaults  0  2
/mnt/backups/default                            /Backups/default        none    bind      0  0
# /home/user/Music
UUID=1dc2e088-3ed0-4ac3-aa23-248f513cddf8       /mnt/Music      ext4    defaults   0 2
/mnt/Music                                      /home/user/Music none    bind       0 0
# 2TB Snapshots for BackInTime, FDE auto-mount
/dev/disk/by-uuid/9e137077-fa14-4386-9424-8e786e4fb866 /mnt/9e137077-fa14-4386-9424-8e786e4fb866 auto nosuid,nodev,nofail,x-gvfs-show 0 0
/mnt/9e137077-fa14-4386-9424-8e786e4fb866 /Backups/backintime none bind 0 0
```
Thanks for all the troubleshooting help so far.

**Comment by Germar** [comment]
I first thought your `bind` mount could make troubles but I double-checked it without errors.

Sorry for the dump question but are you sure, your backups end up on `/Backups/backintime` and not on your root filesystem? What's the path you configured in BiT (`grep 'profile1.snapshots.path=' ~/.config/backintime/config`)?

**Comment by moved-to-gitlab** [comment]
```
user@desktop:~$ grep 'profile1.snapshots.path=' ~/.config/backintime/config
profile1.snapshots.path=/Backups
```
Bind and the mounts and disks are set up this way because I am using a collection of SSDs plus traditional drives, with full-disk encryption on most. My root filesystem is a 256GB drive and the backups definitely won't fit there. They are on an external drive mounted at `/Backups/backintime`.

As a workaround, is there a better way to mount the target drive? It has to be encrypted or this defeats the purpose of using full-disk encryption on the system. It also must automatically mount at boot when I enter my password.

The other backup tool I use, Déjà Dup, encrypts the backup files so I use it with another plain old ext4 drive. That also writes to a subdirectory in `/Backups/` but so far no issues.

**Comment by Germar** [comment]
Ah, now I get a picture of what went wrong. BiT checks free space of the path you configured in Settings (`/Backups` in your case). But the external drive is in `/Backups/backintime` and so it did check free space of `/dev/mapper/ubuntu--vg-root` instead of `/dev/dm-4`.

I'll change it to check the full snapshot path `/Backups/backintime/HOST/USER/ID`

**Comment by Kuphi** [comment]
> Ah, now I get a picture of what went wrong. BiT checks free space of the path you configured in Settings (`/Backups` in your case). But the external drive is in `/Backups/backintime` and so it did check free space of `/dev/mapper/ubuntu--vg-root` instead of `/dev/dm-4`.
> 
> I'll change it to check the full snapshot path `/Backups/backintime/HOST/USER/ID`

I stumpled upon the same or a similar flaw. I set up a new BTRFS drive, encrypted with LUKS. I then changed to owner of /media/user/mountpoint to my user user:user. Everytime backintime creates a new snapshot, any previous backup is deleted. I've deactivated the "remove automatically" feature now and will test again.

Nevertheless, this is not a preferrable outcome. Thank you!
